internal/inventory/app/controller.go

```diff
@@ -40,75 +40,76 @@ const (
 	}
 
-	buff := bytes.Buffer{}
-	decoder := json.NewDecoder(&buff)
-	decoder.UseNumber()
-	decoder.DisallowUnknownFields()
-
 	for i := range event.Readings {
 		reading := &event.Readings[i] // Readings is 169 bytes. This avoid the copy.
-		switch reading.Name {
+		switch reading.ResourceName {
 		default:
 			// this should never happen because it is pre-filtered by the SDK pipeline
-			app.lc.Error("Unknown reading name.", "reading", reading.Name)
+			app.lc.Error("Unknown reading name.", "reading", reading.ResourceName)
 			continue
 
 		case resourceReaderNotification:
-			buff.Reset()
-			buff.WriteString(reading.Value)
 			notification := &llrp.ReaderEventNotification{}
-			if err := decoder.Decode(notification); err != nil {
-				app.lc.Error("Failed to decode reader event notification", "error", err.Error())
+			err := app.getReadingObjectValue(reading.ObjectValue, notification)
+			if err != nil {
+				app.lc.Errorf("Failed to decode reader event notification for device '%s': %s", event.DeviceName, err.Error())
 				continue
 			}
 
-			if err := app.handleReaderEvent(event.Device, notification); err != nil {
+			if err := app.handleReaderEvent(event.DeviceName, notification); err != nil {
 				app.lc.Error("Failed to handle ReaderEventNotification.",
-					"error", err.Error(), "device", event.Device)
+					"error", err.Error(), "device", event.DeviceName)
 			}
 
 		case resourceROAccessReport:
-			buff.Reset()
-			buff.WriteString(reading.Value)
-
 			report := &llrp.ROAccessReport{}
-			if err := decoder.Decode(report); err != nil {
-				app.lc.Error("Failed to decode tag report",
-					"error", err.Error(), "device", event.Device)
+			err := app.getReadingObjectValue(reading.ObjectValue, report)
+			if err != nil {
+				app.lc.Errorf("Failed to decode tag report for device '%s': %s", event.DeviceName, err.Error())
 				continue
 			}
 
 			if report.TagReportData == nil {
-				app.lc.Warn("No tag report data in report.", "device", event.Device)
+				app.lc.Warn("No tag report data in report.", "device", event.DeviceName)
 			} else {
 				// pass the tag report data to the reports channel to be processed by our taskLoop
 				app.reports <- reportData{report, inventory.NewReportInfo(reading)}
 				app.lc.Trace("New ROAccessReport.",
-					"device", event.Device, "tags", len(report.TagReportData))
+					"device", event.DeviceName, "tags", len(report.TagReportData))
 			}
 		}
 	}
 
 	return false, nil
 }
 
+func (app *InventoryApp) getReadingObjectValue(value interface{}, target interface{}) error {
+	// Object reading is of type interface{}, so it gets un-marshaled into a map[string]interface{} when the reading
+	// is un-marshaled by the SDK since the SDK doesn't know the struct. It needs to be re-marshaled back to JSON and
+	// then un-marshaled into the proper target struct that is known by the App Service
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(data, target)
+	return err
+}
```

Comment:
Yup, and thanks for that. I am going to push this into EdgeX so that in Kamakura its a capability of the Reading DTO since this will be a common need.

Reply:
I have refactored this function to use mapstructure

Reply:
I have changed this back to  use json marshal and unmarshal